src/Dict/Any.elm

```diff
@@ -448,7 +483,90 @@ decode fromStr toComparable valueD =
         |> Decode.map (Dict.foldr construct (empty toComparable))
 
 
+{-| Decode a JSON object into an `AnyDict`.
+
+This variant of decode allows you to fail with error while parsing key from String.
+In such case whole Dict decoding will fail.
+
+    import Json.Decode
+
+    type Key = Foo | Bar
+
+    fromString : String -> Result String Key
+    fromString str =
+       case str of
+          "foo" -> Ok Foo
+          "bar" -> Ok Bar
+          _     -> Err <| "Unknown key " ++ str
+
+    toString : Key -> String
+    toString key =
+      case key of
+         Foo -> "foo"
+         Bar -> "bar"
+
+    type alias Data = AnyDict String Key Int
+
+    dataDecoder : Json.Decode.Decoder Data
+    dataDecoder =
+       decode_ (\str _ -> fromString str) toString Json.Decode.int
+
+
+    Json.Decode.decodeString dataDecoder "{\"foo\":1,\"bar\":2}"
+    |> Result.map toList
+    --> Ok [(Bar, 2), (Foo, 1)]
+
+    Json.Decode.decodeString dataDecoder "{\"foo\":1,\"baz\":2}"
+    |> Result.map toList
+    --> Json.Decode.decodeString (Json.Decode.fail "Unknown key baz") "{}"
+
+-}
+decode_ : (String -> v -> Result String k) -> (k -> comparable) -> Decode.Decoder v -> Decode.Decoder (AnyDict comparable k v)
+decode_ fromStr toComparable valueD =
+    let
+        construct strK v =
+            Result.andThen
+                (\acc ->
+                    fromStr strK v
+                        |> Result.map (\key -> insert key v acc)
+                )
+    in
+    Decode.dict valueD
+        |> Decode.map (Dict.foldr construct (Ok <| empty toComparable))
+        |> Decode.andThen
+            (\res ->
+                case res of
+                    Ok val ->
+                        Decode.succeed val
+
+                    Err err ->
+                        Decode.fail err
+            )
```

Comment:
previously this could have been simply decoded to dict with result key and then `andThen`ed like this from `Err` to decoder error. Anyway I think providing this function can make it easier for folks to use this library. Would this be enough for your use case @FranzSkuffka?

Reply:
I understand now that you expect the consumer to pass a function that parses a `String`, rather than passing a `Decode.Decoder key`.

However, this requires me to write more code. My use case is not only matching union types but decoding records for example. There I would have to write `Decode.decodeString myDecoder`. I admit this is just one more line but I do not understand why we use a function with a signature `String -> v -> Result String key` rather than a `Decoder key`. I believe it is more idiomatic and the simpler type signature is a hint that it is actually a simpler solution.


Here is my proposed implementation complete with description and unit tests:
```elm
module AnyDictCodec exposing (..)

import Dict.Any
import Json.Decode as Decode
import Json.Encode as Encode


{-|
    import Dict.Any
    import Json.Decode as Decode
    import Json.Encode as Encode

    type alias Person = {first : String, last : String}
    type alias Age = Int

    personToString : Person -> String
    personToString {first, last} = first ++ last

    personEncode : Person -> Encode.Value
    personEncode {first, last} = Encode.object [("first", (Encode.string first)), ("last", (Encode.string last))]

    personDecode : Decode.Decoder Person
    personDecode =
        Decode.map2
            Person
                (Decode.field "first" Decode.string)
                (Decode.field "last" Decode.string)

    example : Dict.Any.AnyDict String Person Age
    example = Dict.Any.fromList
        personToString [(Person "Jeve" "Sobs", 9001), (Person "Tim" "Berners-Lee", 1234)]

    encode personEncode Encode.int example
        |> Decode.decodeValue (decode personToString personDecode Decode.int)
        --> Ok example
-}
encode : (k -> Encode.Value) -> (v -> Encode.Value) -> Dict.Any.AnyDict comparable k v -> Encode.Value
encode keyEncoder valueEncoder =
    Dict.Any.toList
    >> Encode.list (\(k, v) -> Encode.list identity [keyEncoder k, valueEncoder v])

{-| Decode an AnyDict
-}
decode : (k -> comparable) -> Decode.Decoder k -> (Decode.Decoder v) -> Decode.Decoder (Dict.Any.AnyDict comparable k v)
decode keyToComparable keyDecoder valueDecoder =
    Decode.list (Decode.map2 Tuple.pair (Decode.index 0 keyDecoder) (Decode.index 1 valueDecoder))
    |> Decode.map (Dict.Any.fromList keyToComparable)
```

Reply:
JSON grammar is such that only the string can be a key for JSON. Allowing decoder for any arbitrary JSON value would not be correct. For instance array of Objects can't be a key within object. The same way Int or boolean can't be a key. String parsing is correct approach as the decoder would have unexpected behavior and would allow for completely broken ways of defining decoders.

Your case sounds like you embed stringified JSON into the JSON. In such case you need to call another parsing in json which is embedded in another JSON as string. This library can't support this as a default case because that would mean requiring double quoted strings as a string key (as `"\"key\"": ....`)

Reply:
I am not encoding the dict as an object but as a List of Lists/Tuples. I am aware that keys in JSON always need to be strings.

Reply:
The unit tests / example verification int my proposal pass. That means it produces valid JSON.

Reply:
But the api of the function is misleading. Decoders are for the value type while in this position only string can be present.

Reply:
oh sorry now I see. you propose different json representation. Well I think that's not what users would expect, definitely it's not really compatible with default decoders for regular dict. I understand why one could want to choose this but I think providing decoder for this case is out of the scope of the package itself.

Reply:
Well I consider myself a user of this package and this is exactly what I expect :D I respect your decision either way but I want to stress that in my work I want to persist `AnyDict`s and the approach you proposed makes things more complicated for me. The API you propose in my perspective is only really usable when you have a union type with no arguments as keys.

I think that encoding a JSON as string and using it as a key in an object is a lot more unidiomatic than using a list of tuples.

Reply:
I'm open to discussion here. What about this. Lets keep this topic of alternative json representation separate from documenting existing decoding. Feel free to open PR with alternative approach so we can discuss it. We can of course provide more than just one function to users to pick from based on how they want to represent data in JSON. We just need to make sure the naming makes it clear that this is in fact decoding 2 dimensional list or list of objects like `{ key:...,value: }` or something or perhaps even give a choice to let this up to the user.